tests/main.cpp

```diff
@@ -11,9 +11,19 @@
 #include <doctest/doctest.h>
 #include <vector>
 
+#include <nlohmann/json.hpp>
+
 using namespace std;
 using namespace evm;
 
+TEST_CASE("from_json, to_json tests")
+{
+  Account a1;
+  nlohmann::json j = a1;
+  Account a2 = j.get<Account>();
+  CHECK(a1 == a2);
+}
```

Comment:
Please add tests where each of the fields has been set to a non-default value, so we can be explicit about the expected supported values in both C++ and json.

(ie -
```
Account a3;
a3.address = …;
a3.nonce = …;
… // to json, from json, check equal

nlohmann::json j;
j["address"] = …;
j["nonce"] = …;
… // from json, to json, check equal
```

The example json object should ideally come from elsewhere, either an existing json test case or as written by another Ethereum tool.

Reply:
Agree.

Reply:
If we take half-initialized json object, deserialize it into `Account` and then serialize it into another json, those two are not equal, because `Account` constructor introduces missing fields. Am I missing something? Have a look at last test, it fails.

Reply:
That's correct, `to_json` will always contain all 4 fields so if we compare it against an object containing only 3 of the fields they will never match. For this test we can only `REQUIRE` that each field in the original object is also present in the produced object (and their values match). But we should also have a test that takes a _complete_ json object (including an "address" field) and confirms we can roundtrip it accurately.

Reply:
There is a subtlety here. Addresses are 40 bytes hex strings padded with 0x. As far as I know, it is not restricted to have upper-case hex strings to represent addresses. If we construct an Account from such JSON and then serialize it into another JSON, they may not be equal, because we convert address into lowercase (following etherscan.io and co style).

Reply:
Another subtlety is that some addresses may not be padded with leading 0. So, again, jsons may not be equal.

Reply:
Fair points! There's a mixed-case [checksum address format](https://eips.ethereum.org/EIPS/eip-55) I believed was reasonably standard, but if that's not widely used then doing any string comparison of addresses isn't particularly useful. I like exporting to lower case and fixed 40 bytes. Direct comparison with json objects exported in some other format may not work, but that's ok - the hardcoded test case confirms the _idea_ and in general users should be converting to Account objects and comparing those.